### DISCUSSION TITLE
Chart.js 3 compatibility

### DISCUSSION THREAD

**Discussion by simonbrunel** [chartjs/chartjs-plugin-datalabels#213] [General]
### Status

Chart.js 3 is now fully supported!

Links: [Documentation](https://chartjs-plugin-datalabels.netlify.app/)
Release: [`v2.0.0`](https://github.com/chartjs/chartjs-plugin-datalabels/releases/tag/v2.0.0)
Milestone [2.x](https://github.com/chartjs/chartjs-plugin-datalabels/milestone/9)

### Done

- [x] #233
- [x] Update docs to reflect Chart.js 3 API changes.
- [x] Make the UMD build compatible with the ESM build of Chart.js 3 (#222).
- [x] Provide an ESM build that support tree-shaking for Chart.js helpers (#222).
- [x] Switch to GitHub actions since Travis become unreliable (a8b51cb3d121f0df0673f2f57a0408c589633ddf).

**Comment by Homand-Cindy** [comment]
I followed the procedure, I retrieved the file chartjs-plugin-datalabels.min.js which I call in my mock-up after the file chart.min.js. I display a graph of type sector but the labels are not displayed... did I forget something?

**Reply by alexkuc** [reply]
Can you create a reproducible repository? Or a codepen?

**Reply by Homand-Cindy** [reply]
git clone https://github.com/chartjs/chartjs-plugin-datalabels.git
cd chartjs-plugin-datalabels
git checkout next
npm install
npm run build

and I recovered the file chartjs-plugin-datalabels.min.js in the dist folder

**Reply by alexkuc** [reply]
I meant your own code with the chart where the datalabels are not showing up

**Reply by Homand-Cindy** [reply]
[PieChart.zip](https://github.com/chartjs/chartjs-plugin-datalabels/files/5809742/PieChart.zip)

**Reply by alexkuc** [reply]
Have you seen [Getting Started](https://chartjs-next--chartjs-plugin-datalabels.netlify.app/guide/getting-started.html) section? The plugin does not work as a drop-in approach. After you have included the `.js` file, you still need to configure it. Also, [samples (donut chart)](https://chartjs-next--chartjs-plugin-datalabels.netlify.app/samples/charts/doughnut.html) section is a good place to check to get the hang of how the plugin works.

**Reply by Homand-Cindy** [reply]
thank you for your answer, I had read the startup section but I hadn't read the right version, I had read the production version documentation, sorry

**Comment by saberjsd** [comment]
Hello, It not work for `chart.js@3.0.0-beta.9` when I build `next` branch. When will the latest version be supported for chart.js `master` branch? Many thanks! Here is the error message：
```
TypeError: Cannot read property 'merge' of undefined
(anonymous function)
/chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js:599
  596 |   me._el = el;
  597 | };
  598 | 
> 599 | helpers$1.merge(Label.prototype, {
      | ^  600 |   /**
  601 |    * @private
  602 |    */
```

**Reply by alexkuc** [reply]
How are you building the repository? Can you provide code to reproduce your issue?

**Reply by saberjsd** [reply]
I fork your repository on `next` branch.
```
yarn install
yarn run build
```
[Here is my commit](https://github.com/Saberjsd/chartjs-plugin-datalabels/commit/15ae99904d71d4acba8e643e186468eca0514b9a)

**Reply by alexkuc** [reply]
What is your version of Node? Mine is `v14.15.3`. I was able to build the `next` branch successfully using:

```bash
git clone git@github.com:chartjs/chartjs-plugin-datalabels.git
cd chartjs-plugin-datalabels
git checkout next
npm install
npm run build
```

**Reply by saberjsd** [reply]
`v12.18.4`
Ok,I'll try it right away.

**Reply by alexkuc** [reply]
For the sake of testing I have tried `yarn` as well and it worked on my end:

```bash
git clone git@github.com:chartjs/chartjs-plugin-datalabels.git
cd chartjs-plugin-datalabels
git checkout next
yarn install
yarn run build
```

**Reply by alexkuc** [reply]
So I think it could be due to older version of `Node` (the error you are receiving)

**Reply by saberjsd** [reply]
I can build successfully .The difference is that I use the latest `chart.js@3.0.0-beta.9`

**Reply by saberjsd** [reply]
The build file content is the same when I use `v14.15.3`

**Reply by alexkuc** [reply]
Do you still have an error? I don't follow what's your issue here...

**Reply by saberjsd** [reply]
Yes, [here is this lase build file](https://github.com/Saberjsd/chartjs-plugin-datalabels/tree/chartjs-next/dist)
when I run this code
```javascript
var helpers$1 = chart_js.Chart.helpers;
// chart_js.Chart.helpers  is  undefined
```

**Reply by alexkuc** [reply]
Can you share the code which triggers this error?

**Reply by saberjsd** [reply]
Yeah,
```javascript
import Chart from 'chart.js/dist/chart'

import ChartDataLabels from 'chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js';

// when run here, triggers this error
this.chartInstance = new Chart(canvas, {
  plugins: [ChartDataLabels],
  // other options
});
```

**Reply by alexkuc** [reply]
Can you share a reproducible repo with the source code? It's a little hard to deduct where the error is coming from if all you see is a fraction of the source code.

**Reply by alexkuc** [reply]
Have you also seen [samples section](https://github.com/chartjs/chartjs-plugin-datalabels/tree/chartjs-next/docs/samples)? It contains working examples of different chart types.

**Reply by saberjsd** [reply]
Im sorry to reply you too late beacause I am too busy.
This is my demo: https://github.com/Saberjsd/chart_test/blob/master/src/App.vue
```
yarn install
yarn run serve
```
Look at chrome console can show this error

**Reply by saberjsd** [reply]
Oh! I find the reason! 
The friend @santam85[https://github.com/santam85] is resolve in `chartjs-plugin-datalabels`.But in my project is not resolve.
Can see hash `cefccb5649004401421d89ad096edaeb04e14f76` for more.
Thanks a lot !

**Reply by alexkuc** [reply]
@Saberjsd Hmmm, I get the same error in your repository. I don't really have experience with Vue so I am not able to check if it is a Vue configuration error. In my repositories, this plugin works fine with beta10 of chart.js but I don't use Vue or bundling. My recommendation at this point would be to build a basic example without bundling (use CDN links from npmjs) so that you have a simple hard-coded demo. Then convert that demo into Vue and check if you still have the issues.

**Reply by alexkuc** [reply]
@Saberjsd Look in the [samples section](https://chartjs-plugin-datalabels.netlify.app/samples/) as a starting point

**Reply by saberjsd** [reply]
I build without `dist/chart.esm.js` and push `dist` directory in my forks repository, resolved this problem provisional.
[Here is commit.](https://github.com/Saberjsd/Chart.jsd/commit/0995a57efc0c4dc6e9c44bd4cf6283d20b0b1b80)

**Reply by alexkuc** [reply]
I just realized that this is your issue https://github.com/chartjs/chartjs-plugin-datalabels/pull/222

**Reply by saberjsd** [reply]
Yes, Thank you every much for help me!

**Comment by goonydev** [comment]
Good afternoon,

I don't mean to be rude and I know you guys are probably very busy. But can I get a rough idea of when this plugin is going to be compatible with ChartJs V3? I have developped an application that uses Datalabels and HAS to be running with the latest version of ChartJs. Just before deploying it, ChartJs became incompatible with Datalabels and I have been on stand by ever since. I was under the impression that you guys were close to catch up with ChartJs V3 and kept hanging on an imminent release. Now I just found out that Datalabels V1 is not compatible with ChartJS V3 which I misunderstood would be.

So you guys do what you can, but can you give me an idea? Is it going to take a few weeks? More like 6 months? A year? So I can make decisions about my app.

Thank you.

**Reply by simonbrunel** [reply]
@goonydev About v1 of this plugin, it may take a few years before most users switch to Chart.js v3 and I needed a stable version for maintaining the plugin against Chart.js v2. That's why I released `v1.0.0-beta.1` yesterday but `v2.0.0-beta.1` (for Chart.js v3) will be released very soon (probably today or tomorrow). Though, I can't tell when the final version (`v2.0.0`) will be released since it depends on when Chart.js v3 will be released and how people will test this plugin beta version.

**Reply by goonydev** [reply]
Thanks for the quick response.

I got confused I guess about ChartJS V3 since the CMS my plugin will be used on seems to be already running on ChartJS V3 although it is still a beta version. Thank you for clarifying this to me.

**Comment by snird** [comment]
@simonbrunel and everyone else working on it, you are doing an amazing job.
I upgraded my site to Chart.js 3.2 and use your latest code from `next` branch and it works well.
https://www.the-investing-desk.com/tools/investor-calculator

There's a weird behaviour when you hover that more labels suddenly appear, but otherwise it works as expected.

**Reply by simonbrunel** [reply]
@snird Thank you for your comment. The "weird behavior" your are seeing is probably the same issue as #233.

**Reply by simonbrunel** [reply]
@snird This should be fixed in v2.0.0-rc, can you please confirm?

**Reply by snird** [reply]
@simonbrunel sorry for the late response, I missed the notification.

I can confirm this is now fixed. Thank you,

**Comment by ndkhoa1011** [comment]
Good afternoon,

I am facing a problem with plugin "chartjs-plugin-datalabels (v1.0.0)". Could you teach me how to resolve this error? 
Thanks so much! 

Here is the error message: 

TypeError: Cannot read property 'extend' of undefined
(anonymous function)
node_modules/chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js:587
  584 |   me._el = el;
  585 | };
  586 | 
> 587 | helpers$3.extend(Label.prototype, {
  588 |   /**
  589 |    * @private
  590 |    */

**Reply by alexkuc** [reply]
Can you provide a repository where it would be possible to replicate your error message?

**Reply by ndkhoa1011** [reply]
Thank you for your response!

Here is my repository.

https://github.com/ndkhoa1011/chartjs_tst.git

**Reply by alexkuc** [reply]
Your repository is missing `chart.js` file:

> import { Chart, registerables } from "chart.js";

**Reply by ndkhoa1011** [reply]
I have already defined in package.json.  Please refer the image below.

![image](https://user-images.githubusercontent.com/32250720/116525413-b1073380-a902-11eb-86a8-298127199558.png)

**Reply by alexkuc** [reply]
My bad. I confused CJS with ESM format. The error message looks similar to #222 . @Saberjsd had a similar issue (scroll up) that they managed to resolve. If you are still struggling, have a look in the samples section.

**Reply by ndkhoa1011** [reply]
Thank you for your help!

Let me refer in the samples section.

**Reply by Dishebh** [reply]
Hi, @ndkhoa1011 I am getting the same issue. Did you resolve it?

**Reply by simonbrunel** [reply]
@ndkhoa1011 and @Dishebh you must use version 2.x of this plugin, which is currently in beta:

```sh
npm install -S chartjs-plugin-datalabels@next
```

**Reply by Dishebh** [reply]
@simonbrunel, thanks for the reply. I am getting a TypeeError after doing what you suggested. Please refer to the image attached.  I have tried versions 2.8.0. and 3.2.* of. chart.js and the error is common for both.

<img width="1043" alt="Screenshot 2021-05-04 at 10 13 57 AM" src="https://user-images.githubusercontent.com/43489708/116961685-a57c8980-acc1-11eb-965c-540bceb6c3ee.png">

**Reply by dmahajan980** [reply]
@Dishebh were you able to find a solution to this problem?

**Reply by simonbrunel** [reply]
It's probably because in Chart.js v3, plugins should be registered with `Chart.register(plugin)` ([docs](https://www.chartjs.org/docs/latest/developers/plugins.html#global-plugins)).

**Reply by alexkuc** [reply]
What do you get when you type `Chart.plugins` in console? What about just `Chart`? The error message says that you are trying to use method `.register` on `undefined` variable which suggests that `Chart` is `undefined`.

**Comment by remkovb** [comment]
Hi,

I'm trying to get the RC-2 version of datalabels working with chart.js 3.2.1, but still no success.. It should do something right?

Here is my jsfiddle, could you please point me in the right direction with this?

https://jsfiddle.net/remkovb/zt48huqs/

**Reply by simonbrunel** [reply]
As explained in [the docs](https://v2_0_0-rc--chartjs-plugin-datalabels.netlify.app/guide/migration.html#migrating-to-v2-0-0), you need to explicitly register the plugin. In your case, you need:

```js
Chart.register(ChartDataLabels);
```

https://jsfiddle.net/simonbrunel/nhk8ae03/

**Comment by jipiboily** [comment]
Hello folks!

We're using Chart.js through `ChartjsNodeCanvas` and tried to get the v2 RC1 version of datalabels to work.

None of the plugin loading techniques seems to be working.

I am getting an error like https://github.com/chartjs/chartjs-plugin-datalabels/issues/162

But the workaround suggested in there doesn't seem to apply anymore.

I know it's not the `ChartjsNodeCanvas` repo, but I feel like it might be related to the RC. But I could be wrong :)

So, all that to say, are you aware of potential issues or a way to get it to work?

We can get it to run, once. And then we're getting an error starting at the second call.

How are we loading it? We're registering it with `ChartJS.register(ChartDataLabels);` In `chartCallback`.

But the second call (it's an API end point, no framework, just pure node) throws this:

```javascript
TypeError: Cannot read property 'x' of null
    at orient (/Users/jp/code/grapherate/node_modules/chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js:121:19)
    at Object.fallback [as positioner] (/Users/jp/code/grapherate/node_modules/chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js:345:13)
    at coordinates (/Users/jp/code/grapherate/node_modules/chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js:841:21)
    at Object.draw (/Users/jp/code/grapherate/node_modules/chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js:1012:18)
    at Object.afterDatasetsDraw (/Users/jp/code/grapherate/node_modules/chartjs-plugin-datalabels/dist/chartjs-plugin-datalabels.js:1307:12)
    at callback (/Users/jp/code/grapherate/node_modules/chart.js/dist/chart.js:799:15)
    at PluginService._notify (/Users/jp/code/grapherate/node_modules/chart.js/dist/chart.js:6006:11)
    at PluginService.notify (/Users/jp/code/grapherate/node_modules/chart.js/dist/chart.js:5993:23)
    at Chart.notifyPlugins (/Users/jp/code/grapherate/node_modules/chart.js/dist/chart.js:7057:26)
    at Chart._drawDatasets (/Users/jp/code/grapherate/node_modules/chart.js/dist/chart.js:6833:8)
```

Thoughts anyone?

**Reply by jipiboily** [reply]
PS: I'm happy to pay for an hour of pairing on that issue with someone knowledgable with the internals of all 3 pieces (ChartJS, chartjs-plugin-datalabels and ChartjsNodeCanvas).

**Reply by simonbrunel** [reply]
Can you give us more context: 
- what type of chart are you trying to render (i.e. bar, line, doughnut, financial, etc.)?
- do you use other plugins? if so, which ones and are they working as expected?
- does `Chart.register(ChartDataLabels);` called one time or multiple times per server instance?
- what is `chartCallback`? is it called for every API endpoint calls? can you share this part of the code?
- what Chart.js version are you using exactly?

Are you able to share / setup a repo that we can run to reproduce and debug this issue?

**Reply by jipiboily** [reply]
Sorry for the delayed reply here. It was a holiday here yesterday. I'll set up a repo to reproduce that later today and also provide all the details! :)

**Reply by jipiboily** [reply]
Here is the repro repo: https://github.com/jipiboily/chartjs-node-datalabels-repro

- Trying to render pies here.
- no other plugins
- Chart.js 3.3.0 (I had 3.2.1 in my original repo, but both seems to suffer from the same issue so I didn't force it)

Those lines can be interchanged to test different ways of loading and setting the plugin: https://github.com/jipiboily/chartjs-node-datalabels-repro/blob/ce79a3d94fc23fbe2957e0847ce27fc26c67daba/chartjs_renderer.js#L39:L42

Only line 42 seems to work, but only once. The factory never seems to be called in any case.

I'm not a Node guy (more of a Ruby guy), so I might have screwed up many things (I probably did). It's currently a PoC.

Thanks!

**Reply by simonbrunel** [reply]
The error you get is probably caused by the way `ChartjsNodeCanvas` handles the Chart.js library and plugins internally. Unfortunately, I'm not sure this node integration fully supports Chart.js v3. For example, [`Chart.plugins.register()`](https://github.com/SeanSobey/ChartjsNodeCanvas/blob/master/src/new.ts#L246) is still used while it's not anymore supported by Chart.js v3 (thus you can't use the ["modern" way](https://github.com/SeanSobey/ChartjsNodeCanvas#newer-plugins) to load plugins).

The factory is not called because it's part of the ["legacy" API](https://github.com/SeanSobey/ChartjsNodeCanvas/blob/master/src/legacy.ts#L40) ... but I can't tell you more because I'm not familiar with this project. One way to workaround the issue you are experiencing is to move the construction of the `ChartJSNodeCanvas` outside of the `GraphRenderer` so there is only one instance created for all requests. But I doubt that this solution will be stable, neither will scale.

Please investigate this issue directly in the  `ChartjsNodeCanvas`  repository since you may get better support there. If at the end there is actually a bug in this plugin (and not because of the node integration), feel free to create a new ticket.

**Reply by jipiboily** [reply]
Gotcha. Thanks so much for looking into and the details. This is really appreciated!

**Comment by RaySanjeev** [comment]
i am using chart.js ^3.2.1 and chartjs-plugin-datalabels 2, after importing the chartjs-plugin-datalabels and using it with the plugins key in my chart, I am running my react app but getting an error saying cannot read property skip of undefined. From the source code it is coming from the !el._model.skip. I really need help in fixing this error. Any help on this would be appreciated.

**Reply by RaySanjeev** [reply]
![image](https://user-images.githubusercontent.com/27881244/120082993-9bdd1a80-c0e3-11eb-8e8d-cd0b59384ab9.png)
this is the image of the error

**Reply by RaySanjeev** [reply]
import React, { useEffect, useRef } from "react";
import { Chart } from "react-chartjs-2";
import ChartDataLabels from "chartjs-plugin-datalabels";
// import "chartjs-plugin-labels";

import styles from "./Mixed.module.css";

const MixedChart = (props) => {
	const chartCanvas = useRef(null);

	const data = props.data;

	const spline_x_values = [];
	const scatterDataSpline = [];
	const scatterDataLine = [];
	const pointValues = [];

	if (data[0].items.length > 0) {
		data[0].items.forEach((el) => {
			spline_x_values.push(+el.x_value);
			const obj = {
				x: +el.x_value,
				y: +el.y_value,
				// label: `${el.y_value}`,
			};

			scatterDataSpline.push(obj);
		});
	}

	if (data[1].items.length > 0) {
		data[1].items.forEach((el, i) => {
			const obj = {
				x: spline_x_values[i],
				y: +el.y_value,
				// label: `${el.y_value}`,
			};

			scatterDataLine.push(obj);
		});
	}

	if (data[2].items.length > 0) {
		data[2].items.forEach((el) => {
			const obj = {
				x: +el.x_value,
				y: +el.y_value,
				// label: `${el.y_value}`,
			};

			pointValues.push(obj);
		});
	}

	const configureChart = () => {
		new Chart(chartCanvas.current, {
			type: "scatter",
			data: {
				datasets: [
					{
						label: "Performance",
						data: scatterDataSpline,
						borderColor: "green",
						backgroundColor: "green",
						pointRadius: 5,
						pointHoverRadius: 6,
						tension: 0.4,
						showLine: true,
					},
					{
						label: "Power Goal",
						data: scatterDataLine,
						borderColor: "#1c8bd4",
						backgroundColor: "#1c8bd4",
						pointRadius: 5,
						pointStyle: "circle",
						pointHoverRadius: 6,
						showLine: true,
					},
					{
						label: "Operating Point",
						data: pointValues,
						pointRadius: 7,
						pointHoverRadius: 8,
						pointStyle: "circle",
						borderColor: "red",
						backgroundColor: "red",
						pointBackgroundColor: "red",
						pointBorderColor: "red",
						showLine: true,
					},
				],
			},
			plugins: [ChartDataLabels],
			options: {
				responsive: true,
				hover: {
					mode: "index",
					intersect: true,
				},

				plugins: {
					datalabels: {
						color: "black",
						clamp: true,
						display: function (context) {
							return context.dataset.data[context.dataIndex] >= 1;
						},
					},

					title: {
						display: true,
						text: "Heat Pipe Performance Chart",
						font: {
							size: 20,
						},
						padding: {
							top: 10,
						},
					},
					legend: {
						display: true,
						labels: {
							font: {
								size: 15,
								weight: "bold",
							},
							padding: 30,
						},
					},
				},
				scales: {
					x: {
						beginAtZero: true,
						min: -25,
						// max: 325,
						suggestedMax: 325,
						title: {
							display: true,
							text: "Operating Temperature (C)",
							align: "center",
							padding: 15,
							font: {
								size: 17,
								weight: "bold",
							},
						},
						ticks: {
							font: {
								size: 17,
							},
							stepSize: 25,
						},
					},
					y: {
						beginAtZero: true,
						title: {
							display: true,
							text: "Power (Watts)",
							align: "center",
							padding: 10,
							font: {
								size: 17,
								weight: "bold",
							},
						},
						ticks: {
							font: {
								size: 17,
							},
						},
					},
				},
			},
		});
	};

	useEffect(() => {
		configureChart();
	}, []);

	return (
		<div className={styles.layout}>
			<canvas className={styles.chart} ref={chartCanvas} />
		</div>
	);
};

export default MixedChart;

**Reply by DougDK** [reply]
Seems like you haven't registered ChartDataLabels  
```javascript
Chart.register(ChartDataLabels);
```
or global
```javascript
import Chart from 'chart.js';
import ChartDataLabels from 'chartjs-plugin-datalabels';

// Register the plugin to all charts:
Chart.plugins.register(ChartDataLabels);
```
[chartjs-plugin-datalabels Registration](https://chartjs-plugin-datalabels.netlify.app/guide/getting-started.html#registration)  
[213#discussioncomment-753768](https://github.com/chartjs/chartjs-plugin-datalabels/discussions/213#discussioncomment-753768)

**Comment by DougDK** [comment]
As the #233 seems resolved, is there anything else that I could help with that is blocking the release of this next version? I would really appreciate helping, I have a patch for Chartjs3 in my project but the only blocker is the official release of chartjs-plugin-datalabes@2.0.0

**Reply by DougDK** [reply]
@simonbrunel could you please help me with this?

**Reply by simonbrunel** [reply]
@DougDK [Done](https://github.com/chartjs/chartjs-plugin-datalabels/releases/tag/v2.0.0)!
